File: backend/app/services/cutting_jobs.py
"""Cutting-job request validation and persistence (ticket #94, part of
issue #93's Feature C). No cutting-worker yet — `create_cutting_job` only
ever creates a `queued` job and its `cutting_job_outputs` rows, mirroring
ticket #45's own scoping for AnalysisJob ("data model and request/status
API before the worker exists").
"""

from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.cutting_job import (
    CuttingJob,
    CuttingJobOutput,
    CuttingJobOutputStatus,
    CuttingJobStatus,
)
from app.services.cutting_uploads import (
    UploadNotFoundError,
    get_upload_status,
    source_filename_matches_camera,
    upload_blob_path,
)
from app.services.media_browser import has_cuts
from app.services.media_prober import MediaProbeError, MediaProber
from app.services.s3_client import S3Client, S3ObjectNotFoundError
from app.services.timestamp_excel import (
    ExcelSchemaError,
    InvalidWorkbookError,
    MalformedReferenceCameraError,
    MalformedTestIdError,
    MalformedTimestampCellError,
    TestRowNotFoundError,
    normalize_test_id,
    read_test_row,
)

# One (condition, phase) slot per expected CuttingJobOutput, per uploaded
# camera — ZE_F8 deliberately excluded: every phase is sliced [this
# phase's start, next phase's start), so the last phase (ZE_F8, the 16th
# and final start-only timestamp) has no timestamp to bound its own end and
# can never itself become a Cut (CONTEXT.md's Feature C "ZE_F8 cannot be
# produced" decision, confirmed structurally absent across every `assist`
# branch). Excel's ZE_F8 cell is still parsed (needed to bound ZE_F7's end,
# once a cutting step exists) — it's just never in this list.
_EXPECTED_OUTPUT_SLOTS = [("ME", f"F{n}") for n in range(1, 9)] + [
    ("ZE", f"F{n}") for n in range(1, 8)
]


@dataclass(frozen=True)
class SourceVideoUpload:
    """One uploaded source video, already resolved to a local file — the
    API layer resolves an upload_id (app/services/cutting_uploads.py) into
    this; direct tests of this module construct it from a plain local file,
    with no upload-tracking machinery involved (issue #93's Test seams).
    """

    camera: str
    local_path: Path
    filename: str


class NoSourceVideoUploadedError(Exception):
    """`uploads` is empty — a job needs at least one camera's source video."""


class DuplicateCameraUploadError(Exception):
    """More than one upload claims the same camera."""

    def __init__(self, camera: str) -> None:
        self.camera = camera
        super().__init__(camera)


class InvalidSourceFilenameError(Exception):
    """An uploaded video's filename doesn't loosely match `assist`'s own
    Test/camera convention (re-checked here independent of whatever the
    upload-intake endpoint already validated — same defensive-revalidation
    principle as app/services/analyses.py's `_validate_cut_key`)."""

    def __init__(self, filename: str) -> None:
        self.filename = filename
        super().__init__(filename)


class ReferenceCameraMismatchError(Exception):
    """Exactly one camera was uploaded, and it isn't the Excel row's
    reference camera (CONTEXT.md's Feature C decision: `assist` only
    computes a real audio-sync offset when both cameras are present; with
    only the non-reference camera it would silently default to a zero
    offset instead)."""

    def __init__(self, uploaded_camera: str, reference_camera: str) -> None:
        self.uploaded_camera = uploaded_camera
        self.reference_camera = reference_camera
        super().__init__(uploaded_camera, reference_camera)


class SourceVideoNotDecodableError(Exception):
    """An uploaded video failed ffprobe validation — not a decodable video
    (CONTEXT.md's Feature C decision)."""

    def __init__(self, camera: str) -> None:
        self.camera = camera
        super().__init__(camera)


class SourceVideoCollisionError(Exception):
    """An uploaded source video's derived key already matches an existing
    S3 object (CONTEXT.md's Feature C "S3 collisions" decision: rejected
    outright at upload time, never a silent overwrite)."""

    def __init__(self, key: str) -> None:
        self.key = key
        super().__init__(key)


class CutsAlreadyExistError(Exception):
    """`test_id` already has Cuts under `cuts/<test_id>/` and the caller
    didn't pass `confirm_overwrite` (ticket #96; CONTEXT.md's Feature C "S3
    collisions" decision: re-cutting is allowed, but only once explicitly
    confirmed — unlike `SourceVideoCollisionError`, which is never
    overridable)."""

    def __init__(self, test_id: str) -> None:
        self.test_id = test_id
        super().__init__(test_id)


class CuttingJobNotFoundError(Exception):
    """Raised for a nonexistent CuttingJob id."""


class SourceUploadNotUsableError(Exception):
    """An upload id named in a submission can't be used for it (ticket #97
    moved this check here from the API layer, so one bad upload id fails
    only its own Test within a batch). See the three subclasses."""

    def __init__(self, camera: str) -> None:
        self.camera = camera
        super().__init__(camera)


class UnknownSourceUploadError(SourceUploadNotUsableError):
    """No upload exists with that id."""


class IncompleteSourceUploadError(SourceUploadNotUsableError):
    """The upload exists but hasn't received all its declared bytes yet."""


class MismatchedSourceUploadError(SourceUploadNotUsableError):
    """The upload was declared for a different Test or camera."""


class EmptyBatchError(Exception):
    """A batch submission named no Tests at all."""


class TooManyTestsInBatchError(Exception):
    """A batch submission named more than `MAX_TESTS_PER_BATCH` Tests —
    rejected outright, before any job is created (ticket #97)."""

    def __init__(self, count: int) -> None:
        self.count = count
        super().__init__(count)


class DuplicateTestInBatchError(Exception):
    """A batch submission named the same (normalized) Test more than once —
    rejected outright: two jobs for one Test would race to write the same
    `cuts/<Test>/` keys."""

    def __init__(self, test_id: str) -> None:
        self.test_id = test_id
        super().__init__(test_id)


# Everything one Test's submission can be rejected with. Within a batch,
# any of these fails only that Test (ticket #97); anything else (e.g. a
# database error) is unexpected and propagates.
CUTTING_JOB_VALIDATION_ERRORS: tuple[type[Exception], ...] = (
    SourceUploadNotUsableError,
    NoSourceVideoUploadedError,
    DuplicateCameraUploadError,
    InvalidSourceFilenameError,
    InvalidWorkbookError,
    ExcelSchemaError,
    MalformedTestIdError,
    MalformedReferenceCameraError,
    MalformedTimestampCellError,
    TestRowNotFoundError,
    ReferenceCameraMismatchError,
    SourceVideoNotDecodableError,
    SourceVideoCollisionError,
    CutsAlreadyExistError,
)

# Issue #93's Feature C "Batch submission" decision.
MAX_TESTS_PER_BATCH = 5


def _derive_source_collision_key(test_id: str, filename: str) -> str:
    """The S3-style key an uploaded source video's own filename is checked
    for a collision against.

    Implementation judgment call (CONTEXT.md's Feature C spec names the
    *policy* — "an uploaded source video whose derived filename already
    matches an existing S3 object is rejected outright" — but not the exact
    derivation): reuses the already-validated upload filename verbatim
    under a `source/<test_id>/` namespace, mirroring `cuts/<test_id>/`'s
    existing convention (app/services/media_browser.py's `CUTS_PREFIX`).
    Revisit if a real collision report shows this doesn't match what
    legacy `source/` objects are actually named.
    """
    return f"source/{test_id}/{filename}"


def create_cutting_job(
    db: Session,
    *,
    requested_by_identity: str | None,
    test_id: str,
    excel_bytes: bytes,
    uploads: list[SourceVideoUpload],
    s3: S3Client,
    media_prober: MediaProber,
    confirm_overwrite: bool = False,
) -> CuttingJob:
    """Create a `queued` CuttingJob for `test_id`, attributed to
    `requested_by_identity`, only once every ingestion validation rule from
    issue #93 passes — no partial job is ever created if any check fails.

    `uploads` holds 1-2 already-uploaded source videos (one per camera).
    The Excel's matching row for `test_id` supplies the reference camera
    and phase timestamps; a single-camera upload is only accepted if it
    matches that reference camera. Every uploaded video is probed with
    `media_prober` (rejecting anything not decodable) and checked against
    `s3` for a derived-filename collision before the job (and its expected
    `CuttingJobOutput` rows) is created.

    If `test_id` already has Cuts in S3, raises `CutsAlreadyExistError`
    unless `confirm_overwrite` is set (ticket #96). Checked last, after
    every other rule, so a confirmed follow-up request never fails on
    something the first request could already have reported.
    """
    if not uploads:
        raise NoSourceVideoUploadedError

    # Normalized up front (same as the Excel row's own Test ID cell) so a
    # bare-numeric `test_id` (e.g. "513") matches a well-formed "T513_C1_..."
    # upload filename below, instead of being falsely rejected against the
    # un-normalized form — a real gap found in review.
    test_id = normalize_test_id(test_id)

    seen_cameras: set[str] = set()
    for upload in uploads:
        if upload.camera in seen_cameras:
            raise DuplicateCameraUploadError(upload.camera)
        seen_cameras.add(upload.camera)
        if not source_filename_matches_camera(upload.filename, test_id, upload.camera):
            raise InvalidSourceFilenameError(upload.filename)

    row = read_test_row(excel_bytes, test_id)

    if len(uploads) == 1 and uploads[0].camera != row.reference_camera:
        raise ReferenceCameraMismatchError(uploads[0].camera, row.reference_camera)

    for upload in uploads:
        try:
            media_prober.probe(upload.local_path)
        except MediaProbeError:
            raise SourceVideoNotDecodableError(upload.camera) from None

    for upload in uploads:
        key = _derive_source_collision_key(row.test_id, upload.filename)
        try:
            s3.head_object(key)
        except S3ObjectNotFoundError:
            pass
        else:
            raise SourceVideoCollisionError(key)

    if not confirm_overwrite and has_cuts(s3, row.test_id):
        raise CutsAlreadyExistError(row.test_id)

    job = CuttingJob(
        test_id=row.test_id,
        requested_by_identity=requested_by_identity,
        reference_camera=row.reference_camera,
        phase_timestamps=row.phase_timestamps,
        c1_source_path=next(
            (str(upload.local_path) for upload in uploads if upload.camera == "C1"), None
        ),
        c2_source_path=next(
            (str(upload.local_path) for upload in uploads if upload.camera == "C2"), None
        ),
    )
    job.outputs = [
        CuttingJobOutput(
            camera=upload.camera,
            condition=condition,
            phase=phase,
            status=CuttingJobOutputStatus.PENDING,
        )
        for upload in uploads
        for condition, phase in _EXPECTED_OUTPUT_SLOTS
        if f"{condition}_{phase}" in row.phase_timestamps
    ]

    db.add(job)
    db.commit()
    db.refresh(job)
    return job


@dataclass(frozen=True)
class CuttingJobSubmission:
    """One Test's worth of a submission, naming its source videos by the
    upload ids app/services/cutting_uploads.py minted — not yet resolved or
    validated."""

    test_id: str
    c1_upload_id: str | None = None
    c2_upload_id: str | None = None
    confirm_overwrite: bool = False


@dataclass(frozen=True)
class CuttingJobSubmissionResult:
    """One Test's outcome within a batch: exactly one of `job`/`error` is
    set. `error` is always one of `CUTTING_JOB_VALIDATION_ERRORS`."""

    test_id: str
    job: CuttingJob | None = None
    error: Exception | None = None


def resolve_source_upload(
    upload_root: Path, upload_id: str, *, camera: str, test_id: str
) -> SourceVideoUpload:
    """The local file behind a fully received upload declared for exactly
    this (normalized) `test_id` and `camera`."""
    try:
        upload_status = get_upload_status(upload_root, upload_id)
    except UploadNotFoundError:
        raise UnknownSourceUploadError(camera) from None
    if not upload_status.complete:
        raise IncompleteSourceUploadError(camera)
    if upload_status.test_id != test_id or upload_status.camera != camera:
        raise MismatchedSourceUploadError(camera)
    return SourceVideoUpload(
        camera=camera,
        local_path=upload_blob_path(upload_root, upload_id),
        filename=upload_status.filename,
    )


def submit_cutting_job(
    db: Session,
    *,
    submission: CuttingJobSubmission,
    requested_by_identity: str | None,
    excel_bytes: bytes,
    upload_root: Path,
    s3: S3Client,
    media_prober: MediaProber,
) -> CuttingJob:
    """Resolve `submission`'s upload ids, then `create_cutting_job` — one
    Test, as `POST /cutting-jobs` and each entry of a batch both need."""
    # Normalized the same way `start_upload` already normalized whatever
    # test_id an upload was declared under — otherwise a bare-numeric
    # test_id here would never match a normalized "T..." upload record.
    test_id = normalize_test_id(submission.test_id)
    uploads = [
        resolve_source_upload(upload_root, upload_id, camera=camera, test_id=test_id)
        for camera, upload_id in (("C1", submission.c1_upload_id), ("C2", submission.c2_upload_id))
        if upload_id is not None
    ]
    return create_cutting_job(
        db,
        requested_by_identity=requested_by_identity,
        test_id=test_id,
        excel_bytes=excel_bytes,
        uploads=uploads,
        s3=s3,
        media_prober=media_prober,
        confirm_overwrite=submission.confirm_overwrite,
    )


def submit_cutting_job_batch(
    db: Session,
    *,
    submissions: list[CuttingJobSubmission],
    requested_by_identity: str | None,
    excel_bytes: bytes,
    upload_root: Path,
    s3: S3Client,
    media_prober: MediaProber,
) -> list[CuttingJobSubmissionResult]:
    """Submit up to `MAX_TESTS_PER_BATCH` Tests at once, all read from the
    same timestamp workbook, each becoming its own independent CuttingJob
    (ticket #97). Results come back in submission order.

    An empty or oversized batch, or one naming the same Test twice, is
    rejected outright before any job is created. Past that, each Test
    succeeds or fails on its own: one Test's validation failure is recorded
    in its result and never stops the rest, and each job is committed as
    it's created.
    """
    if not submissions:
        raise EmptyBatchError
    if len(submissions) > MAX_TESTS_PER_BATCH:
        raise TooManyTestsInBatchError(len(submissions))
    test_ids = [normalize_test_id(submission.test_id) for submission in submissions]
    seen_test_ids: set[str] = set()
    for test_id in test_ids:
        if test_id in seen_test_ids:
            raise DuplicateTestInBatchError(test_id)
        seen_test_ids.add(test_id)

    results = []
    for test_id, submission in zip(test_ids, submissions, strict=True):
        try:
            job = submit_cutting_job(
                db,
                submission=submission,
                requested_by_identity=requested_by_identity,
                excel_bytes=excel_bytes,
                upload_root=upload_root,
                s3=s3,
                media_prober=media_prober,
            )
        except CUTTING_JOB_VALIDATION_ERRORS as exc:
            results.append(CuttingJobSubmissionResult(test_id=test_id, error=exc))
        else:
            results.append(CuttingJobSubmissionResult(test_id=job.test_id, job=job))
    return results


def get_cutting_job(db: Session, *, cutting_job_id: int) -> CuttingJob:
    job = db.get(CuttingJob, cutting_job_id)
    if job is None:
        raise CuttingJobNotFoundError(cutting_job_id)
    return job


def claim_next_queued_cutting_job(db: Session) -> CuttingJob | None:
    """Atomically claim the oldest still-`queued` CuttingJob for the
    cutting-worker (ticket #95) to run, or None if the queue is empty.

    `SELECT ... FOR UPDATE SKIP LOCKED`, mirroring `claim_next_queued_job`'s
    own reasoning (app/services/analyses.py) — with exactly one cutting-worker
    instance (CONTEXT.md's Feature C "exactly one CuttingJob running at a
    time platform-wide" decision) this never actually contends, but it's what
    makes the design safe to later run more than one instance without
    double-processing a job.

    Sets `status=running` and `started_at` as part of the same claim, same
    "never observably queued with no owner" reasoning as the analysis
    worker's own claim — there's no per-job version field to record here
    (unlike `dogtrace_version`): a CuttingJob always runs against the
    `assist` snapshot vendored into the cutting-worker image (ticket #112),
    not something worth recording per job.
    """
    job = db.scalar(
        select(CuttingJob)
        .where(CuttingJob.status == CuttingJobStatus.QUEUED)
        .order_by(CuttingJob.created_at.asc())
        .limit(1)
        .with_for_update(skip_locked=True)
    )
    if job is None:
        return None

    job.status = CuttingJobStatus.RUNNING
    job.started_at = datetime.now(UTC)
    db.add(job)
    db.commit()
    db.refresh(job)
    return job


def finalize_cutting_job(db: Session, job: CuttingJob) -> CuttingJob:
    """Set `job`'s terminal status from its outputs' already-recorded final
    statuses. Unlike `finalize_analysis_job`, there's no
    `completed_with_errors` middle ground (CuttingJobStatus only has
    `succeeded`/`failed` as terminal non-cancelled outcomes) — `succeeded`
    only when *every* expected CuttingJobOutput succeeded, `failed`
    otherwise, including a job that errored before any output could be
    attempted (the caller marks every such output `failed` before calling
    this). This all-or-nothing shape is deliberate: the cutting-worker
    (ticket #95) only discards a CuttingJob's local source upload once the
    job reaches `succeeded` (see `orchestrator.py`) — a job left `failed`
    over even one bad phase keeps its source around for a cheap retry,
    rather than forcing a researcher to re-upload a multi-GB video to fix
    one mis-timed phase.

    Call only after any produced Cuts are already durably uploaded to S3.

    `job.outputs` is never expected to be empty in practice (a CuttingJob's
    Excel row needs at least one non-skipped phase to have been created at
    all), but a bare `{...} <= {SUCCEEDED}` is vacuously true for an empty
    set — without the explicit `output_statuses` check below, a job with no
    outputs at all would be marked `succeeded` and have its source upload
    discarded despite producing zero Cuts (caught in review).
    """
    output_statuses = {output.status for output in job.outputs}
    if output_statuses and output_statuses <= {CuttingJobOutputStatus.SUCCEEDED}:
        job.status = CuttingJobStatus.SUCCEEDED
    else:
        job.status = CuttingJobStatus.FAILED

    job.finished_at = datetime.now(UTC)
    db.add(job)
    db.commit()
    db.refresh(job)
    return job


def requeue_stuck_running_cutting_jobs(db: Session) -> list[CuttingJob]:
    """Reset every CuttingJob still `running` back to `queued` (and its
    outputs back to `pending`, clearing any failure_reason), returning the
    jobs that were reset.

    Meant to be called once, before the cutting-worker (ticket #95) starts
    polling, mirroring `requeue_stuck_running_jobs`'s own reasoning — exactly
    one cutting-worker instance ever processes jobs, so a CuttingJob still
    `running` at startup can only mean a previous process crashed or was
    restarted mid-job. Requeuing (not failing) is safe here specifically
    because a CuttingJob's local source upload is only ever discarded on a
    *finalized* `succeeded` job (see `finalize_cutting_job`/
    `orchestrator.py`) — a job stuck `running` never reached that point, so
    its source is still on disk to retry against.
    """
    jobs = list(db.scalars(select(CuttingJob).where(CuttingJob.status == CuttingJobStatus.RUNNING)))
    for job in jobs:
        job.status = CuttingJobStatus.QUEUED
        job.started_at = None
        db.add(job)
        for output in job.outputs:
            output.status = CuttingJobOutputStatus.PENDING
            output.failure_reason = None
            db.add(output)
    if jobs:
        db.commit()
    return jobs

File: frontend/src/services/cuttingJobs.ts
import { API_BASE_URL, errorFromResponse } from '@/services/apiBase'

// Ticket #100, part of issue #93's Feature C: the browser side of Ingestion —
// uploading a Test's C1/C2 source video(s) through the backend's tus-like
// protocol (ticket #94) and submitting up to 5 Tests as one batch (#97).

export type Camera = 'C1' | 'C2'

export interface SourceVideoTarget {
  testId: string
  camera: Camera
}

interface CuttingUploadResponse {
  upload_id: string
  test_id: string
  camera: Camera
  filename: string
  total_size_bytes: number
  received_bytes: number
  complete: boolean
}

// Well under nginx's shared 30 MiB `client_max_body_size`
// (nginx/snippets/server-common.conf), so one chunk is always one request the
// proxy accepts, and a dropped connection costs at most this much re-sent.
export const DEFAULT_UPLOAD_CHUNK_SIZE_BYTES = 8 * 1024 * 1024

// How long to wait before each successive retry after the connection drops
// (or the backend answers 5xx, e.g. mid-restart), before giving up. Reset
// after every chunk that does get through, so a flaky connection only fails
// the upload once it stays down for all of these in a row.
export const DEFAULT_UPLOAD_RETRY_DELAYS_MS = [1_000, 3_000, 10_000, 30_000]

export interface UploadSourceVideoOptions {
  chunkSizeBytes?: number
  retryDelaysMs?: number[]
  // Bytes the backend has confirmed so far, called once up front and again
  // after each chunk.
  onProgress?: (receivedBytes: number, totalBytes: number) => void
}

// A request that never got a usable answer: the connection failed or the
// backend answered 5xx. Worth retrying, unlike a 4xx.
class TransientUploadError extends Error {}

async function uploadRequest(url: string, init: RequestInit, fallback: string): Promise<Response> {
  let response: Response
  try {
    response = await fetch(url, init)
  } catch {
    throw new TransientUploadError('The connection was lost during the upload.')
  }
  if (response.status >= 500) {
    throw new TransientUploadError(fallback)
  }
  return response
}

function uploadUrl(uploadId: string): string {
  return `${API_BASE_URL}/cutting-jobs/uploads/${encodeURIComponent(uploadId)}`
}

// A lost response here can leave an unused, empty upload behind on the
// backend once retried. That's harmless: an empty upload holds no storage
// against the cap, which counts received bytes.
async function startUpload(file: File, target: SourceVideoTarget): Promise<CuttingUploadResponse> {
  const response = await uploadRequest(
    `${API_BASE_URL}/cutting-jobs/uploads`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        test_id: target.testId,
        camera: target.camera,
        filename: file.name,
        total_size_bytes: file.size,
      }),
    },
    'Failed to start the upload.',
  )
  if (!response.ok) {
    throw await errorFromResponse(response, 'Failed to start the upload.')
  }
  return response.json()
}

async function appendChunk(
  uploadId: string,
  offset: number,
  chunk: Blob,
): Promise<CuttingUploadResponse> {
  const fallback = 'Failed to upload the video.'
  const response = await uploadRequest(
    uploadUrl(uploadId),
    { method: 'PATCH', headers: { 'Upload-Offset': String(offset) }, body: chunk },
    fallback,
  )
  // 409: our offset is stale (a dropped request may still have written part
  // of its chunk) or the upload is already complete. Either way the backend's
  // own status says where to carry on from.
  if (response.status === 409) {
    return getUpload(uploadId)
  }
  if (!response.ok) {
    throw await errorFromResponse(response, fallback)
  }
  return response.json()
}

async function getUpload(uploadId: string): Promise<CuttingUploadResponse> {
  const upload = await findUpload(uploadId)
  if (upload === null) {
    throw new Error('The upload was removed from the server. Start it again.')
  }
  return upload
}

// `null` when the backend no longer has `uploadId` (e.g. it was consumed by a
// job that has since succeeded and discarded its source).
async function findUpload(uploadId: string): Promise<CuttingUploadResponse | null> {
  const fallback = 'Failed to check the upload.'
  const response = await uploadRequest(uploadUrl(uploadId), {}, fallback)
  if (response.status === 404) return null
  if (!response.ok) {
    throw await errorFromResponse(response, fallback)
  }
  return response.json()
}

// Which upload id each file went to, so uploading the same file again — after
// giving up on a dropped connection, or after a page reload — continues that
// upload rather than sending every byte again. Kept in localStorage, with an
// in-memory copy for when storage is unavailable (private window, blocked
// site data); losing it only costs a fresh upload, never correctness, since
// the backend's own status is always re-read before resuming.
const REMEMBERED_UPLOAD_PREFIX = 'cutting-upload:'
const rememberedInMemory = new Map<string, string>()

function rememberedUploadKey(file: File, target: SourceVideoTarget): string {
  return (
    REMEMBERED_UPLOAD_PREFIX +
    JSON.stringify([
      target.testId.trim().toUpperCase(),
      target.camera,
      file.name,
      file.size,
      file.lastModified,
    ])
  )
}

function recallUpload(key: string): string | null {
  try {
    const stored = localStorage.getItem(key)
    if (stored !== null) return stored
  } catch {
    // Storage unavailable: fall back to this page's own memory.
  }
  return rememberedInMemory.get(key) ?? null
}

function rememberUpload(key: string, uploadId: string): void {
  rememberedInMemory.set(key, uploadId)
  try {
    localStorage.setItem(key, uploadId)
  } catch {
    // Storage unavailable: the in-memory copy still covers this page.
  }
}

// Stop resuming `file`'s upload for `target`: call once a job has been
// created from it, so a later re-cut uploads the file afresh instead of
// pointing a second job at the first one's source.
export function forgetSourceVideoUpload(file: File, target: SourceVideoTarget): void {
  const key = rememberedUploadKey(file, target)
  rememberedInMemory.delete(key)
  try {
    localStorage.removeItem(key)
  } catch {
    // Storage unavailable: nothing stored there to remove.
  }
}

// The backend's upload for `file`, continuing a remembered one when the
// backend still has it for exactly this file, else a new one.
async function resumeOrStartUpload(
  file: File,
  target: SourceVideoTarget,
): Promise<CuttingUploadResponse> {
  const key = rememberedUploadKey(file, target)
  const rememberedId = recallUpload(key)
  if (rememberedId !== null) {
    const existing = await findUpload(rememberedId)
    if (
      existing !== null &&
      existing.camera === target.camera &&
      existing.filename === file.name &&
      existing.total_size_bytes === file.size
    ) {
      return existing
    }
  }
  const upload = await startUpload(file, target)
  rememberUpload(key, upload.upload_id)
  return upload
}

function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

// Uploads `file` as `target`'s source video and resolves to its upload id,
// once the backend has every byte. A dropped connection is resumed from the
// byte count the backend reports, never restarted from zero.
export async function uploadSourceVideo(
  file: File,
  target: SourceVideoTarget,
  options: UploadSourceVideoOptions = {},
): Promise<string> {
  const chunkSize = options.chunkSizeBytes ?? DEFAULT_UPLOAD_CHUNK_SIZE_BYTES
  const retryDelays = options.retryDelaysMs ?? DEFAULT_UPLOAD_RETRY_DELAYS_MS

  let upload: CuttingUploadResponse | null = null
  let failedAttempts = 0
  for (;;) {
    try {
      if (upload === null) {
        upload = await resumeOrStartUpload(file, target)
      } else if (failedAttempts > 0) {
        // After a failure, the backend may hold more than we last saw: part
        // of a dropped chunk can still have been written.
        upload = await getUpload(upload.upload_id)
      }
      options.onProgress?.(upload.received_bytes, upload.total_size_bytes)
      if (upload.complete) return upload.upload_id

      const offset = upload.received_bytes
      upload = await appendChunk(upload.upload_id, offset, file.slice(offset, offset + chunkSize))
      failedAttempts = 0
    } catch (err) {
      if (!(err instanceof TransientUploadError) || failedAttempts >= retryDelays.length) {
        throw err
      }
      await wait(retryDelays[failedAttempts]!)
      failedAttempts++
    }
  }
}

// Mirrors the backend's CuttingJobStatus / CuttingJobOutputStatus enums
// (backend/app/models/cutting_job.py).
export type CuttingJobStatus = 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled'
export type CuttingJobOutputStatus = 'pending' | 'succeeded' | 'failed'

export const CUTTING_JOB_STATUS_LABELS: Record<CuttingJobStatus, string> = {
  queued: 'Queued',
  running: 'Running',
  succeeded: 'Succeeded',
  failed: 'Failed',
  cancelled: 'Cancelled',
}

// Same limit as the backend's MAX_TESTS_PER_BATCH (app/services/cutting_jobs.py),
// which enforces it regardless; this only stops the form from offering more.
export const MAX_TESTS_PER_BATCH = 5

// One expected Cut: a (camera, condition, phase) the job will produce.
export interface CuttingJobOutput {
  camera: Camera
  condition: string
  phase: string
  status: CuttingJobOutputStatus
  failureReason: string | null
}

export interface CuttingJob {
  id: number
  testId: string
  requestedByIdentity: string | null
  status: CuttingJobStatus
  referenceCamera: Camera
  createdAt: string
  startedAt: string | null
  finishedAt: string | null
  outputs: CuttingJobOutput[]
}

interface CuttingJobOutputResponse {
  camera: Camera
  condition: string
  phase: string
  status: CuttingJobOutputStatus
  failure_reason: string | null
}

interface CuttingJobResponse {
  id: number
  test_id: string
  requested_by_identity: string | null
  status: CuttingJobStatus
  reference_camera: Camera
  created_at: string
  started_at: string | null
  finished_at: string | null
  outputs: CuttingJobOutputResponse[]
}

function toCuttingJob(row: CuttingJobResponse): CuttingJob {
  return {
    id: row.id,
    testId: row.test_id,
    requestedByIdentity: row.requested_by_identity,
    status: row.status,
    referenceCamera: row.reference_camera,
    createdAt: row.created_at,
    startedAt: row.started_at,
    finishedAt: row.finished_at,
    outputs: row.outputs.map((output) => ({
      camera: output.camera,
      condition: output.condition,
      phase: output.phase,
      status: output.status,
      failureReason: output.failure_reason,
    })),
  }
}

export interface CuttingJobBatchEntry {
  testId: string
  c1UploadId: string | null
  c2UploadId: string | null
}

// Ticket #96's marker on a "Cuts already exist" rejection (the backend's
// CUTS_ALREADY_EXIST_CODE), as opposed to every other per-Test error.
export const CUTS_ALREADY_EXIST_CODE = 'cuts_already_exist'

// Why one Test got no job. `code` is CUTS_ALREADY_EXIST_CODE when the Test
// already has Cuts and the submission didn't confirm overwriting them;
// `detail` is always safe to show as-is.
export interface CuttingJobBatchError {
  detail: string
  code: string | null
}

// Exactly one of `job`/`error` is set. `testId` is the backend's normalized
// form (`513` comes back as `T513`).
export interface CuttingJobBatchResult {
  testId: string
  job: CuttingJob | null
  error: CuttingJobBatchError | null
}

interface CuttingJobBatchResponse {
  results: {
    test_id: string
    job: CuttingJobResponse | null
    error: { status_code: number; detail: string; code: string | null } | null
  }[]
}

// Creates one independent CuttingJob per Test (up to MAX_TESTS_PER_BATCH),
// all read from the one timestamp workbook. Resolves with one result per
// entry, in the same order, as soon as the batch was accepted — a Test failing
// its own validation is a result, not a thrown error. Throws only when the
// batch as a whole was refused (too many Tests, the same Test twice, the
// workbook too large, throttled), with the backend's own message.
export async function submitCuttingJobBatch(
  excel: File,
  entries: CuttingJobBatchEntry[],
): Promise<CuttingJobBatchResult[]> {
  const formData = new FormData()
  formData.append('excel', excel)
  formData.append(
    'tests',
    JSON.stringify(
      entries.map((entry) => ({
        test_id: entry.testId,
        c1_upload_id: entry.c1UploadId,
        c2_upload_id: entry.c2UploadId,
      })),
    ),
  )

  const response = await fetch(`${API_BASE_URL}/cutting-jobs/batch`, {
    method: 'POST',
    body: formData,
  })
  if (!response.ok) {
    throw await errorFromResponse(response, 'Failed to submit the cutting jobs.')
  }

  const body: CuttingJobBatchResponse = await response.json()
  return body.results.map((result) => ({
    testId: result.test_id,
    job: result.job === null ? null : toCuttingJob(result.job),
    error: result.error === null ? null : { detail: result.error.detail, code: result.error.code },
  }))
}

export async function getCuttingJob(id: number): Promise<CuttingJob> {
  const response = await fetch(`${API_BASE_URL}/cutting-jobs/${id}`)
  if (!response.ok) {
    throw await errorFromResponse(response, 'Failed to load the cutting job.')
  }
  return toCuttingJob(await response.json())
}
